File: examples/haitun-workspace/skills/fusion-flow-next/src/planning.ts
import type { Diagnostic } from "./types.js";

/** A DSL syntax capability required by one planned function. */
export interface PlannedSyntax {
  readonly description: string;
  /**
   * `null` explicitly means the planner could not find matching DSL syntax; it
   * must not invent a name. A non-null name must be non-empty after trimming.
   * Empty or whitespace-only names are invalid, and the future checker treats
   * them as unmapped.
   */
  readonly name: string | null;
}

export interface PlannedFunction {
  readonly id: string;
  readonly description: string;
  /**
   * Must contain at least one mapping. Empty means invalid or incomplete
   * planner output; the future checker warns and sets `canAuthorWorkflow` to
   * false.
   */
  readonly syntax: readonly PlannedSyntax[];
}

export interface PlanningCheckResult {
  /**
   * Means only that the declared planned functions have acceptable syntax
   * mappings. It does not imply DSL parse validity, backend lowering, or
   * successful execution.
   */
  readonly canAuthorWorkflow: boolean;
  readonly diagnostics: readonly Diagnostic[];
}

/**
 * Haitun must call this after listing planned functions and before authoring the
 * DSL.
 *
 * The caller supplies the syntax names that are actually available; a
 * non-empty name is not assumed to exist. A future implementation emits
 * warnings for null or unavailable syntax mappings and sets
 * `canAuthorWorkflow` to false. It only checks declared items and cannot prove
 * that the LLM listed every required function. The current implementation
 * throws only because this module is not wired or implemented; a future
 * missing mapping is a normal result, not an exception.
 */
export function checkPlannedFunctions(
  functions: readonly PlannedFunction[],
  availableSyntaxNames: readonly string[],
): PlanningCheckResult {
  void functions;
  void availableSyntaxNames;
  throw new Error("FusionFlow Next planning check is not implemented.");
}

File: examples/haitun-workspace/skills/fusion-flow-next/fusion_flow_next/planning.py
"""Validate planned DSL capabilities before workflow authoring."""

from __future__ import annotations

from dataclasses import dataclass

from .contracts import Diagnostic


@dataclass(frozen=True, slots=True)
class PlannedSyntax:
    """DSL syntax required by a planned function.

    ``name=None`` means no matching syntax was found; callers must not invent
    one. Non-null names must be non-empty after trimming.
    """

    description: str
    name: str | None


@dataclass(frozen=True, slots=True)
class PlannedFunction:
    """Planned function with at least one required syntax mapping."""

    id: str
    description: str
    syntax: tuple[PlannedSyntax, ...]


@dataclass(frozen=True, slots=True)
class PlanningCheckResult:
    """Whether declared functions can be authored, independent of later phases."""

    can_author_workflow: bool
    diagnostics: tuple[Diagnostic, ...]


def check_planned_functions(
    functions: tuple[PlannedFunction, ...],
    available_syntax_names: tuple[str, ...],
) -> PlanningCheckResult:
    """Check planned functions after planning and before authoring the DSL.

    The caller supplies syntax names that actually exist; a non-empty mapping
    is not assumed to be available. Missing, blank, unavailable, or empty
    mappings are normal diagnostics and make ``can_author_workflow`` false.
    This phase checks declared items only and cannot prove the planner listed
    every required function. It does not imply parse, compile, or execution
    success.

    The current stub raises only because this phase is not implemented.
    """

    del functions, available_syntax_names
    raise NotImplementedError("FusionFlow Next planning check is not implemented.")
